File: engine/src/tutorial/TutorialSystem.cpp
#include "tutorial/TutorialSystem.h"

namespace subspace {

// ---------------------------------------------------------------------------
// TutorialStep
// ---------------------------------------------------------------------------

void TutorialStep::Start() {
    status = TutorialStepStatus::Active;
    elapsedTime = 0.0f;
}

void TutorialStep::Complete() {
    status = TutorialStepStatus::Completed;
}

void TutorialStep::Skip() {
    status = TutorialStepStatus::Skipped;
}

void TutorialStep::Reset() {
    status = TutorialStepStatus::NotStarted;
    elapsedTime = 0.0f;
}

bool TutorialStep::IsTimeElapsed() const {
    return type == TutorialStepType::WaitForTime && elapsedTime >= duration;
}

// ---------------------------------------------------------------------------
// Tutorial
// ---------------------------------------------------------------------------

bool Tutorial::Start() {
    if (status != TutorialStatus::NotStarted) return false;
    status = TutorialStatus::Active;
    currentStepIndex = 0;
    if (!steps.empty()) {
        steps[0].Start();
    }
    return true;
}

bool Tutorial::CompleteCurrentStep() {
    if (status != TutorialStatus::Active) return false;
    if (currentStepIndex < 0 || currentStepIndex >= static_cast<int>(steps.size()))
        return false;

    steps[currentStepIndex].Complete();

    int nextIndex = currentStepIndex + 1;
    if (nextIndex >= static_cast<int>(steps.size())) {
        Complete();
    } else {
        currentStepIndex = nextIndex;
        steps[currentStepIndex].Start();
    }
    return true;
}

void Tutorial::SkipCurrentStep() {
    if (status != TutorialStatus::Active) return;
    if (currentStepIndex < 0 || currentStepIndex >= static_cast<int>(steps.size()))
        return;

    TutorialStep& step = steps[currentStepIndex];
    if (!step.canSkip) return;

    step.Skip();

    int nextIndex = currentStepIndex + 1;
    if (nextIndex >= static_cast<int>(steps.size())) {
        Complete();
    } else {
        currentStepIndex = nextIndex;
        steps[currentStepIndex].Start();
    }
}

void Tutorial::Skip() {
    status = TutorialStatus::Skipped;
    for (auto& step : steps) {
        if (step.status == TutorialStepStatus::NotStarted ||
            step.status == TutorialStepStatus::Active) {
            step.Skip();
        }
    }
}

void Tutorial::Complete() {
    status = TutorialStatus::Completed;
}

void Tutorial::Reset() {
    status = TutorialStatus::NotStarted;
    currentStepIndex = 0;
    for (auto& step : steps) {
        step.Reset();
    }
}

void Tutorial::Update(float deltaTime) {
    if (status != TutorialStatus::Active) return;
    if (currentStepIndex < 0 || currentStepIndex >= static_cast<int>(steps.size()))
        return;

    TutorialStep& step = steps[currentStepIndex];
    if (step.status == TutorialStepStatus::Active &&
        step.type == TutorialStepType::WaitForTime) {
        step.elapsedTime += deltaTime;
        if (step.IsTimeElapsed()) {
            CompleteCurrentStep();
        }
    }
}

TutorialStep* Tutorial::GetCurrentStep() {
    if (currentStepIndex < 0 || currentStepIndex >= static_cast<int>(steps.size()))
        return nullptr;
    return &steps[currentStepIndex];
}

float Tutorial::GetCompletionPercentage() const {
    if (steps.empty()) return 0.0f;
    int completed = 0;
    for (const auto& step : steps) {
        if (step.status == TutorialStepStatus::Completed ||
            step.status == TutorialStepStatus::Skipped) {
            ++completed;
        }
    }
    return static_cast<float>(completed) / static_cast<float>(steps.size()) * 100.0f;
}

bool Tutorial::AreAllStepsComplete() const {
    for (const auto& step : steps) {
        if (step.status != TutorialStepStatus::Completed &&
            step.status != TutorialStepStatus::Skipped) {
            return false;
        }
    }
    return true;
}

// ---------------------------------------------------------------------------
// TutorialComponent serialization
// ---------------------------------------------------------------------------

static std::string TutorialStatusToString(TutorialStatus s) {
    switch (s) {
    case TutorialStatus::NotStarted: return "NotStarted";
    case TutorialStatus::Active:     return "Active";
    case TutorialStatus::Completed:  return "Completed";
    case TutorialStatus::Skipped:    return "Skipped";
    }
    return "NotStarted";
}

static TutorialStatus TutorialStatusFromString(const std::string& s) {
    if (s == "Active")    return TutorialStatus::Active;
    if (s == "Completed") return TutorialStatus::Completed;
    if (s == "Skipped")   return TutorialStatus::Skipped;
    return TutorialStatus::NotStarted;
}

static std::string StepTypeToString(TutorialStepType t) {
    switch (t) {
    case TutorialStepType::Message:       return "Message";
    case TutorialStepType::WaitForKey:    return "WaitForKey";
    case TutorialStepType::WaitForAction: return "WaitForAction";
    case TutorialStepType::HighlightUI:   return "HighlightUI";
    case TutorialStepType::WaitForTime:   return "WaitForTime";
    }
    return "Message";
}

static TutorialStepType StepTypeFromString(const std::string& s) {
    if (s == "WaitForKey")    return TutorialStepType::WaitForKey;
    if (s == "WaitForAction") return TutorialStepType::WaitForAction;
    if (s == "HighlightUI")   return TutorialStepType::HighlightUI;
    if (s == "WaitForTime")   return TutorialStepType::WaitForTime;
    return TutorialStepType::Message;
}

static std::string StepStatusToString(TutorialStepStatus s) {
    switch (s) {
    case TutorialStepStatus::NotStarted: return "NotStarted";
    case TutorialStepStatus::Active:     return "Active";
    case TutorialStepStatus::Completed:  return "Completed";
    case TutorialStepStatus::Skipped:    return "Skipped";
    }
    return "NotStarted";
}

static TutorialStepStatus StepStatusFromString(const std::string& s) {
    if (s == "Active")    return TutorialStepStatus::Active;
    if (s == "Completed") return TutorialStepStatus::Completed;
    if (s == "Skipped")   return TutorialStepStatus::Skipped;
    return TutorialStepStatus::NotStarted;
}

ComponentData TutorialComponent::Serialize() const {
    ComponentData cd;
    cd.componentType = "TutorialComponent";

    // Active tutorials
    cd.data["activeTutorialCount"] = std::to_string(activeTutorials.size());
    for (size_t i = 0; i < activeTutorials.size(); ++i) {
        const auto& tut = activeTutorials[i];
        std::string prefix = "tut_" + std::to_string(i) + "_";
        cd.data[prefix + "id"] = tut.id;
        cd.data[prefix + "title"] = tut.title;
        cd.data[prefix + "status"] = TutorialStatusToString(tut.status);
        cd.data[prefix + "currentStep"] = std::to_string(tut.currentStepIndex);
        cd.data[prefix + "autoStart"] = tut.autoStart ? "true" : "false";

        cd.data[prefix + "stepCount"] = std::to_string(tut.steps.size());
        for (size_t j = 0; j < tut.steps.size(); ++j) {
            const auto& step = tut.steps[j];
            std::string sp = prefix + "step_" + std::to_string(j) + "_";
            cd.data[sp + "id"] = step.id;
            cd.data[sp + "type"] = StepTypeToString(step.type);
            cd.data[sp + "status"] = StepStatusToString(step.status);
            cd.data[sp + "requiredAction"] = step.requiredAction;
            cd.data[sp + "canSkip"] = step.canSkip ? "true" : "false";
        }
    }

    // Completed tutorial ids
    cd.data["completedCount"] = std::to_string(completedTutorialIds.size());
    int idx = 0;
    for (const auto& cid : completedTutorialIds) {
        cd.data["completed_" + std::to_string(idx)] = cid;
        ++idx;
    }

    return cd;
}

void TutorialComponent::Deserialize(const ComponentData& data) {
    activeTutorials.clear();
    completedTutorialIds.clear();

    auto getStr = [&](const std::string& key) -> std::string {
        auto it = data.data.find(key);
        return it != data.data.end() ? it->second : "";
    };
    auto getInt = [&](const std::string& key, int def = 0) -> int {
        auto it = data.data.find(key);
        if (it == data.data.end()) return def;
        try { return std::stoi(it->second); } catch (...) { return def; }
    };

    // Active tutorials
    int tutCount = getInt("activeTutorialCount", 0);
    for (int i = 0; i < tutCount; ++i) {
        std::string prefix = "tut_" + std::to_string(i) + "_";
        Tutorial tut;
        tut.id = getStr(prefix + "id");
        tut.title = getStr(prefix + "title");
        tut.status = TutorialStatusFromString(getStr(prefix + "status"));
        tut.currentStepIndex = getInt(prefix + "currentStep", 0);
        tut.autoStart = getStr(prefix + "autoStart") == "true";

        int stepCount = getInt(prefix + "stepCount", 0);
        for (int j = 0; j < stepCount; ++j) {
            std::string sp = prefix + "step_" + std::to_string(j) + "_";
            TutorialStep step;
            step.id = getStr(sp + "id");
            step.type = StepTypeFromString(getStr(sp + "type"));
            step.status = StepStatusFromString(getStr(sp + "status"));
            step.requiredAction = getStr(sp + "requiredAction");
            step.canSkip = getStr(sp + "canSkip") != "false";
            tut.steps.push_back(std::move(step));
        }

        activeTutorials.push_back(std::move(tut));
    }

    // Completed tutorials
    int completedCount = getInt("completedCount", 0);
    for (int i = 0; i < completedCount; ++i) {
        std::string cid = getStr("completed_" + std::to_string(i));
        if (!cid.empty()) {
            completedTutorialIds.insert(cid);
        }
    }
}

// ---------------------------------------------------------------------------
// TutorialSystem
// ---------------------------------------------------------------------------

TutorialSystem::TutorialSystem() : SystemBase("TutorialSystem") {}

void TutorialSystem::Update(float /*deltaTime*/) {
    // Standalone: callers drive per-entity tutorials via the public API.
}

void TutorialSystem::AddTutorialTemplate(const Tutorial& tutorial) {
    _templates[tutorial.id] = tutorial;
}

bool TutorialSystem::StartTutorial(EntityId /*entityId*/,
                                   const std::string& templateId,
                                   TutorialComponent& comp) {
    auto it = _templates.find(templateId);
    if (it == _templates.end()) return false;

    if (!ArePrerequisitesMet(comp, it->second.prerequisites)) return false;

    // Don't start if already active or completed
    if (comp.completedTutorialIds.count(templateId)) return false;
    for (const auto& tut : comp.activeTutorials) {
        if (tut.id == templateId) return false;
    }

    Tutorial instance = it->second;
    instance.Start();
    comp.activeTutorials.push_back(std::move(instance));
    return true;
}

void TutorialSystem::CompleteCurrentStep(TutorialComponent& comp,
                                         const std::string& tutorialId) {
    // When tutorialId is empty, complete the first active tutorial's current step only.
    for (auto& tut : comp.activeTutorials) {
        if (tutorialId.empty() || tut.id == tutorialId) {
            if (tut.status == TutorialStatus::Active) {
                tut.CompleteCurrentStep();
                if (tut.status == TutorialStatus::Completed) {
                    comp.completedTutorialIds.insert(tut.id);
                }
                return;
            }
        }
    }
}

void TutorialSystem::CompleteActionStep(TutorialComponent& comp,
                                        const std::string& action) {
    // Intentionally completes all active tutorials with a matching WaitForAction step.
    for (auto& tut : comp.activeTutorials) {
        if (tut.status != TutorialStatus::Active) continue;
        TutorialStep* step = tut.GetCurrentStep();
        if (step && step->status == TutorialStepStatus::Active &&
            step->type == TutorialStepType::WaitForAction &&
            step->requiredAction == action) {
            tut.CompleteCurrentStep();
            if (tut.status == TutorialStatus::Completed) {
                comp.completedTutorialIds.insert(tut.id);
            }
        }
    }
}

void TutorialSystem::SkipTutorial(TutorialComponent& comp,
                                  const std::string& tutorialId) {
    for (auto& tut : comp.activeTutorials) {
        if (tut.id == tutorialId) {
            tut.Skip();
            comp.completedTutorialIds.insert(tut.id);
            return;
        }
    }
}

bool TutorialSystem::HasCompletedTutorial(const TutorialComponent& comp,
                                          const std::string& tutorialId) const {
    return comp.completedTutorialIds.count(tutorialId) > 0;
}

bool TutorialSystem::ArePrerequisitesMet(
    const TutorialComponent& comp,
    const std::vector<std::string>& prereqs) const {
    for (const auto& prereq : prereqs) {
        if (comp.completedTutorialIds.count(prereq) == 0) return false;
    }
    return true;
}

void TutorialSystem::CheckAutoStartTutorials(EntityId entityId,
                                             TutorialComponent& comp) {
    for (const auto& [templateId, tmpl] : _templates) {
        if (!tmpl.autoStart) continue;
        if (comp.completedTutorialIds.count(templateId)) continue;

        // Skip if already active
        bool alreadyActive = false;
        for (const auto& tut : comp.activeTutorials) {
            if (tut.id == templateId) {
                alreadyActive = true;
                break;
            }
        }
        if (alreadyActive) continue;

        if (ArePrerequisitesMet(comp, tmpl.prerequisites)) {
            StartTutorial(entityId, templateId, comp);
        }
    }
}

const std::unordered_map<std::string, Tutorial>&
TutorialSystem::GetTutorialTemplates() const {
    return _templates;
}

size_t TutorialSystem::GetTemplateCount() const { return _templates.size(); }

} // namespace subspace

File: AvorionLike/Core/Tutorial/TutorialSystem.cs
using AvorionLike.Core.ECS;
using AvorionLike.Core.Events;
using AvorionLike.Core.Logging;

namespace AvorionLike.Core.Tutorial;

/// <summary>
/// System that manages tutorial progression and events
/// </summary>
public class TutorialSystem : SystemBase
{
    private readonly EntityManager _entityManager;
    private readonly EventSystem _eventSystem;
    private static readonly HashSet<string> EmptyCompletedSet = new();
    
    /// <summary>
    /// All tutorial templates available in the game
    /// </summary>
    private readonly Dictionary<string, Tutorial> _tutorialTemplates = new();
    
    /// <summary>
    /// Active tutorials for each entity
    /// </summary>
    private readonly Dictionary<Guid, List<Tutorial>> _activeTutorials = new();
    
    /// <summary>
    /// Completed tutorial IDs for each entity
    /// </summary>
    private readonly Dictionary<Guid, HashSet<string>> _completedTutorials = new();
    
    /// <summary>
    /// Create a new tutorial system
    /// </summary>
    /// <param name="entityManager">Entity manager</param>
    /// <param name="eventSystem">Event system</param>
    public TutorialSystem(EntityManager entityManager, EventSystem eventSystem) : base("TutorialSystem")
    {
        _entityManager = entityManager ?? throw new ArgumentNullException(nameof(entityManager));
        _eventSystem = eventSystem ?? throw new ArgumentNullException(nameof(eventSystem));
        
        // Subscribe to game events to track player actions
        _eventSystem.Subscribe(GameEvents.EntityCreated, OnEntityCreated);
        _eventSystem.Subscribe(GameEvents.ResourceCollected, OnResourceCollected);
        _eventSystem.Subscribe("EntityDestroyed", OnEntityDestroyed);
        _eventSystem.Subscribe("QuestAccepted", OnQuestAccepted);
    }
    
    /// <summary>
    /// Update all active tutorials
    /// </summary>
    /// <param name="deltaTime">Time elapsed since last update in seconds</param>
    public override void Update(float deltaTime)
    {
        foreach (var kvp in _activeTutorials.ToList())
        {
            var entityId = kvp.Key;
            var tutorials = kvp.Value;
            
            for (int i = tutorials.Count - 1; i >= 0; i--)
            {
                var tutorial = tutorials[i];
                var previousStatus = tutorial.Status;
                
                tutorial.Update();
                
                // Check if tutorial status changed
                if (tutorial.Status != previousStatus)
                {
                    HandleTutorialStatusChange(entityId, tutorial, previousStatus);
                    
                    // Remove completed/skipped tutorials
                    if (tutorial.Status == TutorialStatus.Completed || 
                        tutorial.Status == TutorialStatus.Skipped)
                    {
                        tutorials.RemoveAt(i);
                        
                        // Track completed tutorials
                        if (!_completedTutorials.ContainsKey(entityId))
                        {
                            _completedTutorials[entityId] = new HashSet<string>();
                        }
                        _completedTutorials[entityId].Add(tutorial.Id);
                        
                        // Check for tutorials that can now be started
                        CheckAutoStartTutorials(entityId);
                    }
                }
            }
        }
    }
    
    /// <summary>
    /// Handle tutorial status changes
    /// </summary>
    private void HandleTutorialStatusChange(Guid entityId, Tutorial tutorial, TutorialStatus previousStatus)
    {
        switch (tutorial.Status)
        {
            case TutorialStatus.Active:
                Logger.Instance.Info("TutorialSystem", $"Tutorial '{tutorial.Title}' started for entity {entityId}");
                _eventSystem.Publish("TutorialStarted", new TutorialEvent
                {
                    EntityId = entityId,
                    TutorialId = tutorial.Id,
                    TutorialTitle = tutorial.Title
                });
                break;
                
            case TutorialStatus.Completed:
                Logger.Instance.Info("TutorialSystem", $"Tutorial '{tutorial.Title}' completed by entity {entityId}");
                _eventSystem.Publish("TutorialCompleted", new TutorialEvent
                {
                    EntityId = entityId,
                    TutorialId = tutorial.Id,
                    TutorialTitle = tutorial.Title
                });
                break;
                
            case TutorialStatus.Skipped:
                Logger.Instance.Info("TutorialSystem", $"Tutorial '{tutorial.Title}' skipped by entity {entityId}");
                _eventSystem.Publish("TutorialSkipped", new TutorialEvent
                {
                    EntityId = entityId,
                    TutorialId = tutorial.Id,
                    TutorialTitle = tutorial.Title
                });
                break;
        }
    }
    
    /// <summary>
    /// Add a tutorial template
    /// </summary>
    /// <param name="tutorial">Tutorial template</param>
    public void AddTutorialTemplate(Tutorial tutorial)
    {
        if (_tutorialTemplates.ContainsKey(tutorial.Id))
        {
            Logger.Instance.Warning("TutorialSystem", $"Tutorial template '{tutorial.Id}' already exists, replacing");
        }
        
        _tutorialTemplates[tutorial.Id] = tutorial;
        Logger.Instance.Info("TutorialSystem", $"Added tutorial template: {tutorial.Title} ({tutorial.Id})");
    }
    
    /// <summary>
    /// Start a tutorial for an entity
    /// </summary>
    /// <param name="entityId">Entity to start tutorial for</param>
    /// <param name="tutorialId">Tutorial template ID</param>
    /// <returns>True if tutorial was started</returns>
    public bool StartTutorial(Guid entityId, string tutorialId)
    {
        if (!_tutorialTemplates.TryGetValue(tutorialId, out var template))
        {
            Logger.Instance.Warning("TutorialSystem", $"Tutorial template '{tutorialId}' not found");
            return false;
        }
        
        // Check if already completed
        if (_completedTutorials.TryGetValue(entityId, out var completed) && 
            completed.Contains(tutorialId))
        {
            Logger.Instance.Info("TutorialSystem", $"Tutorial '{tutorialId}' already completed by entity {entityId}");
            return false;
        }
        
        // Check prerequisites
        if (!ArePrerequisitesMet(entityId, template.Prerequisites))
        {
            Logger.Instance.Warning("TutorialSystem", $"Prerequisites not met for tutorial '{tutorialId}'");
            return false;
        }
        
        // Create instance from template
        var tutorial = CloneTutorial(template);
        
        if (!_activeTutorials.ContainsKey(entityId))
        {
            _activeTutorials[entityId] = new List<Tutorial>();
        }
        
        _activeTutorials[entityId].Add(tutorial);
        tutorial.Start();
        
        return true;
    }
    
    /// <summary>
    /// Complete the current step of a tutorial
    /// </summary>
    /// <param name="entityId">Entity with the tutorial</param>
    /// <param name="tutorialId">Tutorial ID (optional, completes first active if null)</param>
    public void CompleteCurrentStep(Guid entityId, string? tutorialId = null)
    {
        if (!_activeTutorials.TryGetValue(entityId, out var tutorials))
            return;
            
        Tutorial? targetTutorial = tutorialId != null 
            ? tutorials.FirstOrDefault(t => t.Id == tutorialId)
            : tutorials.FirstOrDefault(t => t.Status == TutorialStatus.Active);
            
        if (targetTutorial != null)
        {
            targetTutorial.CompleteCurrentStep();
            
            _eventSystem.Publish("TutorialStepCompleted", new TutorialStepEvent
            {
                EntityId = entityId,
                TutorialId = targetTutorial.Id,
                StepTitle = targetTutorial.CurrentStep?.Title ?? "Unknown"
            });
        }
    }
    
    /// <summary>
    /// Skip a tutorial
    /// </summary>
    /// <param name="entityId">Entity with the tutorial</param>
    /// <param name="tutorialId">Tutorial ID</param>
    public void SkipTutorial(Guid entityId, string tutorialId)
    {
        if (!_activeTutorials.TryGetValue(entityId, out var tutorials))
            return;
            
        var tutorial = tutorials.FirstOrDefault(t => t.Id == tutorialId);
        if (tutorial != null)
        {
            tutorial.Skip();
        }
    }
    
    /// <summary>
    /// Get active tutorials for an entity
    /// </summary>
    /// <param name="entityId">Entity ID</param>
    /// <returns>List of active tutorials</returns>
    public List<Tutorial> GetActiveTutorials(Guid entityId)
    {
        return _activeTutorials.TryGetValue(entityId, out var tutorials) 
            ? tutorials 
            : new List<Tutorial>();
    }
    
    /// <summary>
    /// Check if an entity has completed a tutorial
    /// </summary>
    /// <param name="entityId">Entity ID</param>
    /// <param name="tutorialId">Tutorial ID</param>
    /// <returns>True if completed</returns>
    public bool HasCompletedTutorial(Guid entityId, string tutorialId)
    {
        return _completedTutorials.TryGetValue(entityId, out var completed) && 
               completed.Contains(tutorialId);
    }
    
    /// <summary>
    /// Check if prerequisites are met
    /// </summary>
    private bool ArePrerequisitesMet(Guid entityId, List<string> prerequisites)
    {
        if (prerequisites.Count == 0)
            return true;
            
        if (!_completedTutorials.TryGetValue(entityId, out var completed))
            return false;
            
        return prerequisites.All(prereq => completed.Contains(prereq));
    }
    
    /// <summary>
    /// Check for auto-start tutorials
    /// </summary>
    private void CheckAutoStartTutorials(Guid entityId)
    {
        foreach (var template in _tutorialTemplates.Values.Where(t => t.AutoStart))
        {
            if (ArePrerequisitesMet(entityId, template.Prerequisites))
            {
                StartTutorial(entityId, template.Id);
            }
        }
    }
    
    /// <summary>
    /// Clone a tutorial template
    /// </summary>
    private Tutorial CloneTutorial(Tutorial template)
    {
        var tutorial = new Tutorial
        {
            Id = template.Id, // Keep template ID for tracking
            Title = template.Title,
            Description = template.Description,
            AutoStart = template.AutoStart,
            Prerequisites = new List<string>(template.Prerequisites)
        };
        
        foreach (var stepTemplate in template.Steps)
        {
            tutorial.Steps.Add(new TutorialStep
            {
                Type = stepTemplate.Type,
                Title = stepTemplate.Title,
                Message = stepTemplate.Message,
                RequiredKey = stepTemplate.RequiredKey,
                RequiredAction = stepTemplate.RequiredAction,
                UIElementId = stepTemplate.UIElementId,
                Duration = stepTemplate.Duration,
                CanSkip = stepTemplate.CanSkip
            });
        }
        
        return tutorial;
    }
    
    // Event handlers
    
    private void OnEntityCreated(GameEvent eventData)
    {
        if (eventData is EntityEvent entityEvent)
        {
            CheckAutoStartTutorials(entityEvent.EntityId);
        }
    }
    
    private void OnResourceCollected(GameEvent eventData)
    {
        if (eventData is ResourceEvent resourceEvent)
        {
            // Progress any tutorials waiting for resource collection
            CompleteActionBasedSteps(resourceEvent.EntityId, "collect_resource");
        }
    }
    
    private void OnEntityDestroyed(GameEvent eventData)
    {
        if (eventData is EntityEvent entityEvent)
        {
            // Progress any tutorials waiting for combat
            var allEntities = _activeTutorials.Keys.ToList();
            foreach (var entityId in allEntities)
            {
                CompleteActionBasedSteps(entityId, "destroy_entity");
            }
        }
    }
    
    private void OnQuestAccepted(GameEvent eventData)
    {
        // Quest system events can trigger tutorial progression
        CompleteActionBasedSteps(Guid.Empty, "accept_quest");
    }
    
    /// <summary>
    /// Complete steps waiting for a specific action
    /// </summary>
    private void CompleteActionBasedSteps(Guid entityId, string action)
    {
        if (!_activeTutorials.TryGetValue(entityId, out var tutorials))
            return;
            
        foreach (var tutorial in tutorials.Where(t => t.Status == TutorialStatus.Active))
        {
            var currentStep = tutorial.CurrentStep;
            if (currentStep != null &&
                currentStep.Type == TutorialStepType.WaitForAction &&
                currentStep.RequiredAction == action &&
                currentStep.Status == TutorialStepStatus.Active)
            {
                tutorial.CompleteCurrentStep();
            }
        }
    }
    
    /// <summary>
    /// Get all tutorial templates
    /// </summary>
    /// <returns>All tutorial templates</returns>
    public IReadOnlyDictionary<string, Tutorial> GetTutorialTemplates()
    {
        return _tutorialTemplates;
    }

    /// <summary>
    /// Snapshot the system state into a TutorialComponent for persistence.
    /// Call this before saving the game.
    /// </summary>
    /// <param name="entityId">Entity whose tutorial state to capture</param>
    /// <returns>A populated TutorialComponent, or null if no state exists</returns>
    public TutorialComponent? CaptureState(Guid entityId)
    {
        var hasActive = _activeTutorials.TryGetValue(entityId, out var active);
        var hasCompleted = _completedTutorials.TryGetValue(entityId, out var completed);

        if (!hasActive && !hasCompleted)
            return null;

        var component = new TutorialComponent
        {
            EntityId = entityId,
            ActiveTutorials = hasActive ? new List<Tutorial>(active!) : new List<Tutorial>(),
            CompletedTutorialIds = hasCompleted ? new HashSet<string>(completed!) : new HashSet<string>()
        };

        return component;
    }

    /// <summary>
    /// Restore tutorial state from a previously saved TutorialComponent.
    /// Call this after loading a game.
    /// </summary>
    /// <param name="component">The deserialized TutorialComponent</param>
    public void RestoreState(TutorialComponent component)
    {
        var entityId = component.EntityId;

        if (component.ActiveTutorials.Count > 0)
        {
            _activeTutorials[entityId] = new List<Tutorial>(component.ActiveTutorials);
        }

        if (component.CompletedTutorialIds.Count > 0)
        {
            _completedTutorials[entityId] = new HashSet<string>(component.CompletedTutorialIds);
        }
    }

    /// <summary>
    /// Get completed tutorial IDs for an entity
    /// </summary>
    /// <param name="entityId">Entity ID</param>
    /// <returns>Set of completed tutorial IDs</returns>
    public IReadOnlySet<string> GetCompletedTutorialIds(Guid entityId)
    {
        return _completedTutorials.TryGetValue(entityId, out var completed)
            ? completed
            : EmptyCompletedSet;
    }
}

/// <summary>
/// Event data for tutorial-related events
/// </summary>
public class TutorialEvent : GameEvent
{
    public Guid EntityId { get; set; }
    public string TutorialId { get; set; } = string.Empty;
    public string TutorialTitle { get; set; } = string.Empty;
}

/// <summary>
/// Event data for tutorial step events
/// </summary>
public class TutorialStepEvent : GameEvent
{
    public Guid EntityId { get; set; }
    public string TutorialId { get; set; } = string.Empty;
    public string StepTitle { get; set; } = string.Empty;
}
